### PR TITLE
Make TlsTransport and associated classes extensible outside the library

### PR DESCRIPTION
**Pull request by rajeevmv**
Make TlsTransport and associated classes extensible outside the library

### CODE REVIEW

**Comment by rajeevmv** [comment]
@arturl,

The changes I am proposing are to make TlsTransport extensible on the server-side. It should not affect UWP client support.

**Comment by arturl** [comment]
Sounds good, thanks. No concerns from me,

File: Microsoft.Azure.Amqp/Amqp/Transport/TlsTransport.cs [hunk 1]
```diff
@@ -5,19 +5,20 @@ namespace Microsoft.Azure.Amqp.Transport
...
         static readonly AsyncCallback onOpenComplete = OnOpenComplete;
         static readonly AsyncCallback onWriteComplete = OnWriteComplete;
         static readonly AsyncCallback onReadComplete = OnReadComplete;
         const SslProtocols DefaultSslProtocols = SslProtocols.Tls12 | SslProtocols.Tls11 | SslProtocols.Tls;
-        readonly TransportBase innerTransport;
-        readonly CustomSslStream sslStream;
-        TlsTransportSettings tlsSettings;
+        protected readonly TransportBase innerTransport;
+        protected readonly CustomSslStream sslStream;
+        protected TlsTransportSettings tlsSettings;
```
Comment (xinchen10) [inline Microsoft.Azure.Amqp/Amqp/Transport/TlsTransport.cs]: why the above three need to be protected?

File: Microsoft.Azure.Amqp/Amqp/Transport/TlsTransportInitiator.cs [hunk 2]
```diff
@@ -8,11 +8,11 @@ namespace Microsoft.Azure.Amqp.Transport
     /// <summary>
     /// This initiator establishes an SSL connection (no AMQP security upgrade)
     /// </summary>
-    sealed class TlsTransportInitiator : TransportInitiator
+    public class TlsTransportInitiator : TransportInitiator
     {
         static readonly AsyncCallback onTransportOpened = OnTransportOpened;
 
-        TlsTransportSettings transportSettings;
+        protected TlsTransportSettings transportSettings;
```
Comment (xinchen10) [inline Microsoft.Azure.Amqp/Amqp/Transport/TlsTransportInitiator.cs]: If this is needed in derived classes, add a protected getter.

File: Microsoft.Azure.Amqp/Amqp/Transport/TlsTransportInitiator.cs [hunk 3]
```diff
@@ -47,6 +47,11 @@ public override bool ConnectAsync(TimeSpan timeout, TransportAsyncCallbackArgs c
             }
         }
 
+        protected virtual TlsTransport OnCreateInnerTransport(TransportAsyncCallbackArgs innerArgs)
```
Comment (xinchen10) [inline Microsoft.Azure.Amqp/Amqp/Transport/TlsTransportInitiator.cs]: This should be named "OnCreateTransport". It creates a new one from an inner transport. If you pass transport settings in the method like below, do you still the transportSettings to be protected?
protected virtual TlsTransport OnCreateTransport(TransportBase innerTransport, TlsTransportSettings settings).

File: Microsoft.Azure.Amqp/Amqp/Transport/TlsTransportListener.cs [hunk 4]
```diff
@@ -8,11 +8,11 @@ namespace Microsoft.Azure.Amqp.Transport
...
     /// This listener accepts SSL transport directly (no AMQP security upgrade)
     /// </summary>
-    sealed class TlsTransportListener : TransportListener
+    public class TlsTransportListener : TransportListener
     {
         readonly AsyncCallback onTransportOpened;
-        readonly TlsTransportSettings transportSettings;
-        TransportListener innerListener;
+        protected readonly TlsTransportSettings transportSettings;
+        protected TransportListener innerListener;
```
Comment (xinchen10) [inline Microsoft.Azure.Amqp/Amqp/Transport/TlsTransportListener.cs]: why do you need access to innerListener?

File: Microsoft.Azure.Amqp/Amqp/Transport/TlsTransportListener.cs [hunk 5]
```diff
@@ -47,6 +47,11 @@ protected override void OnListen()
             this.innerListener.Listen(this.OnAcceptInnerTransport);
         }
 
+        protected virtual TlsTransport OnCreateInnerTransport(TransportAsyncCallbackArgs innerArgs)
```
Comment (xinchen10) [inline Microsoft.Azure.Amqp/Amqp/Transport/TlsTransportListener.cs]: Same comment as in TlsTransportInitiator

File: Microsoft.Azure.Amqp/Amqp/Transport/TlsTransport.cs [hunk 6]
```diff
@@ -154,6 +146,19 @@ protected override void AbortInternal()
...
 
+        protected virtual X509Principal CreateX509Principal()
+        {
+            // Cannot cast from X509Certificate to X509Certificate2
+            // using workaround mentioned here: https://github.com/dotnet/corefx/issues/4510
+            var cert = new X509Certificate2(this.sslStream.RemoteCertificate.Export(X509ContentType.Cert));
+            return new X509Principal(new X509CertificateIdentity(cert, this.sslStream.IsRemoteCertificateValid));
+        }
+
+        protected virtual bool RemoteCertificateValidationCallback(object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors)
```
Comment (xinchen10) [inline Microsoft.Azure.Amqp/Amqp/Transport/TlsTransport.cs]: rename this to "ValidateRemoteCertificate"

File: Microsoft.Azure.Amqp/Amqp/Transport/TlsTransport.cs [hunk 7]
```diff
@@ -154,6 +146,19 @@ protected override void AbortInternal()
             this.innerTransport.Abort();
         }
 
+        protected virtual X509Principal CreateX509Principal()
```
Comment (xinchen10) [inline Microsoft.Azure.Amqp/Amqp/Transport/TlsTransport.cs]: Can we pass the cert and other info as parameters in this method?

**Comment by rajeevmv** [comment]
Closing this one in favor of another against develop branch

**Pull request closed without merging**